### DISCUSSION TITLE
User Model Flow Chart

### DISCUSSION THREAD

**Discussion by ernest-phillips** [ernest-phillips/home_inventory#9] [Ideas]
This is my initial design of the `User` model.

```mermaid
flowchart TD
    A[User]
    C[Residence_1] 
    D[Residence_2]
    E[Items_1]
    F[Image_1]
    G[Items_2]
    H[Image_2]
    I[Image_3]
    J[Items_3]

    A --> |has many|C 
    A --> |has many|D
    C --> |has many|E --> |Has One|F 
    D --> |has many|G --> |Has One|H 
    D --> |has many|J --> |Has One|I 
```

Essentially, in a User account, you can create inventory lists for multiple residences. There is one sole user per account, at least in my first iteration. There may be need for allowing a secondary user to access the inventory from the same account.
Each residence would have its own list of items within it. Each of those items can have one image of the item. 

A user can add, delete or edit item entries.
They can also add, delete or edit residences. 

Here's a table showing the primary models and their attributes.

|   User	|   Residence	|  Item 	|   	|
|---	|---	|---	|---	|
|  Name 	|   Address	|  Name 	|   	|
|  Email	|   	|  Description 	|   	|
|  Password 	|   	|  Quantity 	|   	|
|   	|   	|  Type	|   	|
|   	|   	|  Room 	|   	|
|   	|   	|  Purchase Amount 	|   	|
|   	|   	|  Model 	|   	|
|   	|   	|  Serial Number 	|   	|
|   	|   	|  Purchase Date 	|   	|
|   	|   	|  Value 	|   	|
|   	|   	|  Image {via association} 	|   	|

Not sure if I have too much or too little here.

**Comment by ernest-phillips** [comment]
I added a `password` attribute to the user model for authorization.
I removed the `residences` from the chart, because the individual `Residence` is what will be associated with the `User`

**Comment by francisfuzz** [comment]
Questions:

- Dependent destroys: If a user is deleted, do their residences and images get deleted too, or not?
- On editing residences: is the only editable property the address? Otherwise, is it possible for a user to "give" a residence to another user? 
- Item schema: what are the datatypes for each field?

**Reply by ernest-phillips** [reply]
> Dependent destroys: If a user is deleted, do their residences and images get deleted too, or not?

I was actually considering a Soft Delete/Destroy action for database records, whereby the records appear to be deleted byt a database record is retained. 
https://entrision.com/blog/comparing-paranoia-vs-discard/

> On editing residences: is the only editable property the address? Otherwise, is it possible for a user to "give" a residence to another user? 

I think the idea of transferring a residence to someone else is a possible feature to consider in the future. For the first iteration, I plan to only have a sole owner of these residences. At the time of building this initial draft out , I couldn't think of another relevant data point for the residence itself. 
However, the following possibilities come to mind:
`owner`  : string
`financed`: string
`insurance_provider` : string
`paid_off` : bool 
`price` : decimal
`household_members` : string

Here's the other data types for the `item` schema.
https://github.com/ernest-phillips/home-inventory/blob/0291ad16e7ebedea7c8bc7f3342ffa4b8e88daac/hi-backend/db/migrate/20230403213742_create_items.rb#L4-L10
